kiarash/JS/JS-generators/index.js

```diff
@@ -0,0 +1,36 @@
+import all from "it-all";
+
+async function* delayedRandomSource() {
+  let max = 5;
+  let min = 0;
+
+  while (true) {
+    let randomDelay = Math.floor(Math.random() * (max - min + 1)) + min;
+    let randomNumber = Math.random();
+
+    if (randomNumber > 0.9) {
+      return;
+    }
+
+    yield new Promise((resolve) => {
+      setTimeout(() => {
+        resolve(randomNumber);
+      }, randomDelay * 1000);
+    });
+  }
+}
+
+let generatorRunner = async (generatorFunction) => {
+  let generator = generatorFunction();
```

Comment:
The result of calling a generator function is called an _iterator_, because it's a data structure that _iterates_ over the generated results.